beacon_node/beacon_chain/src/eth1_chain.rs

```diff
@@ -231,83 +222,54 @@ impl<T: EthSpec, S: Store<T>> Eth1ChainBackend<T> for CachingEth1Backend<T, S> {
 
         let blocks = self.core.blocks().read();
 
-        let (new_eth1_data, all_eth1_data) = if let Some(sets) = eth1_data_sets(
-            blocks.iter(),
-            prev_eth1_hash,
-            voting_period_start_seconds,
-            spec,
-            &self.log,
-        ) {
-            sets
-        } else {
-            // The algorithm was unable to find the `new_eth1_data` and `all_eth1_data` sets.
-            //
-            // This is likely because the caches are empty or the previous eth1 block hash is not
-            // in the cache.
-            //
-            // This situation can also be caused when a testnet does not have an adequate delay
-            // between the eth1 genesis block and the eth2 genesis block. This delay needs to be at
-            // least `2 * ETH1_FOLLOW_DISTANCE`.
-            crit!(
-                self.log,
-                "Unable to find eth1 data sets";
-                "lowest_block_number" => self.core.lowest_block_number(),
-                "earliest_block_timestamp" => self.core.earliest_block_timestamp(),
-                "genesis_time" => state.genesis_time,
-                "outcome" => "casting random eth1 vote"
-            );
-
-            return Ok(random_eth1_data());
-        };
+        let votes_to_consider =
+            get_votes_to_consider(blocks.iter(), voting_period_start_seconds, spec);
 
         trace!(
             self.log,
-            "Found eth1 data sets";
-            "all_eth1_data" => all_eth1_data.len(),
-            "new_eth1_data" => new_eth1_data.len(),
+            "Found eth1 data votes_to_consider";
+            "votes_to_consider" => votes_to_consider.len(),
         );
-
-        let valid_votes = collect_valid_votes(state, new_eth1_data, all_eth1_data);
+        let valid_votes = collect_valid_votes(state, &votes_to_consider);
 
         let eth1_data = if let Some(eth1_data) = find_winning_vote(valid_votes) {
             eth1_data
         } else {
-            // In this case, there are no other viable votes (perhaps there are no votes yet or all
-            // the existing votes are junk).
+            // In this case, there are no valid votes available.
             //
-            // Here we choose the latest block in our voting window.
-            blocks
+            // Here we choose the eth1_data corresponding to the latest block in our voting window.
+            // If no votes exist, choose `state.eth1_data` as default vote.
+            let default_vote = votes_to_consider
                 .iter()
-                .rev()
-                .skip_while(|eth1_block| eth1_block.timestamp > voting_period_start_seconds)
-                .skip(eth1_follow_distance as usize)
-                .next()
-                .map(|block| {
-                    trace!(
+                .max_by(|(_, x), (_, y)| x.cmp(y))
+                .map(|vote| {
+                    let vote = vote.0.clone();
+                    warn!(
```

Comment:
I don't think this should be a `warn`, maybe `debug` at most. This is expected to happen on the first block of every Eth1 voting period, when the list of `state.eth1_data_votes` is empty.

Reply:
Ohh yaa. Fixing!